doc/cli_wrapper/shd-cli_wrapper_hld.md

```diff
@@ -0,0 +1,213 @@
+
+Everytime user executes a command, CLI engine will load main source file (ex. config/main.py), then starts to execute the command. In this process, all libraries always have to be imported again. It costs about 80% execution time on this process, so CLI_wrapper is designed to reduce the time.
+
+The figure below depicts the architecture of CLI_wrapper. 
+
+![Architecture](images/architecture_1.jpg)
+
+* CLI_wrapper builder
+
+  In order to integrate CLI_wrapper with current CLI architecture, we develop CLI_wrapper builder to generate CLI_wrapper for command source file (ex. main.py). When sonic-utilities debian is packaged, it calls CLI_wrapper builder to generate wrapped files which contains only function interface and minimal libraries. It replaces function body with the calling of CLI_wrapper client to send a encoded command to CLI_wrapper server and acquires the result. Since the CLI_wrapper only contains minimal libraries, CLI engine doesn't need a lot of time to execute.
+
+![Architecture](images/architecture_2.jpg)
+
+* CLI_wrapper server
+
+  CLI_wrapper server is responsible for the execution of click command. It starts when system boots up, and loads all needed libraries to server. Then setup a socket to wait client's request. Once it accepts a request, it spawns a CLI_worker process to deal with this request. 
+  
+  In order to avoid that some command may stuck in CLI worker and cause resource leakage, a watchdog thread is designed to watch the execution time of CLI worker, once the execution time exceeds maximum execution time, the CLI worker will be gracefully terminated. But if the CLI worker is still alive after one minute, a kill signal will be sent to CLI worker to tear down it. 
+
+  The local DB is used to store the record of spawned CLI worker, it records the steady timestamp and pid of the CLI worker. 
+
+* CLI_worker
+
+  Decode the request and execute this command. Return the execution result to client and finish this process.
+
+  Everytime cli_wapper server receives a request, it spawns a process to handle this request, so no single block point may happen in this design. 
+
+* CLI_wrapper client
+
+  It is a library that can be used to interact with CLI wrapper server. 
+
+---
+
+## 1.3 Design
+
+This section provides design details for CLI_wrapper.
+
+### 1.3.1 Generate wrapped files
+
+![Generate wrapper file](images/generate_wrapper_file.jpg)
+
+* In the step 1 and step 2, builder copies the necessary libraries to wrapped file and discard other libraries to reduce loading time
+
+* In the step 3, builder removes the content of command function body and replaces it with CLI_wrapper client
+
+### 1.3.2 Command execution flow
+
+![Command execution flow](images/command_execution_flow.jpg)
+
+* System ctrl bring up CLI_wrapper server when system starts

```

Comment:
Please keep this feature disabled by default, if we're comfortable making it enabled some point in time in the future, we can have it enabled by default.

Reply:
Do we have an option to use current click commands even after this feature is enabled? I just want to make sure we're not blocked because of some bug in the wrapper routines?